### ISSUE TITLE
systemd-boot: loaderentry.install hook lacks the ability to detect cryptdevice option, making boot recovery imposible

### ISSUE THREAD

**Issue by eugenesan** [systemd/systemd#32147]
### systemd version the issue has been seen with

255.4

### Used distribution

Ubuntu 24.04

### Linux kernel version used

kernel 6.8.0-22

### CPU architectures issue was seen on

None

### Component

systemd-boot

### Expected behaviour you didn't see

It's impossible to generate working boot entries (eg. missing cryptdevice options) when loaderentry.install hook is invoked in a chroot. Since it assumes correct option already exists in cmdline.
loaderentry.install should be able to detect required cryptdevice options similarly to grub hooks.

### Unexpected behavior you saw

If current cmdline lacks proper cryptdevice option (for example when trying to recover from LiveCD in chroot) generated entry lacks required to boot options.

### Steps to reproduce the problem

1. Boot LiveCD/Recovery media
2. Mount encrypted root/boot volume
3. Trigger operation invoking loaderentry.install hook (for example update-initramfs -u)
4. Observe cryptdevice option missing from generated boot entry

**Comment by poettering** [comment]
I am not grokking what you are writing here, but if you need to configure a LUKS device to boot off, you can do so via /etc/kernel/cmdline?

**Comment by eugenesan** [comment]
@poettering 
Sorry, I went overboard with "words" trying to follow "New Issue" guidelines...

TLDR;
I was expecting sd-boot hooks to detect kernel options required to mount root filesystem but instead after reading the code I realized it just extracts relevant part from already configured .*/.*/cmdline.
This critical when trying to recover from boot failure using chroot from other system or doing initial system installation.

We never had to manually edit cmdline to generate boot entries with grub.

**Comment by YHNdnzj** [comment]
Note that the needed options solely depend on the initrd generator. IOW, `cryptdevice=` is not a thing in systemd at all. If the initrd generator uses systemd in initrd, e.g. mkosi-initrd, mkinitcpio with systemd hook, or dracut, `rd.luks.*` options. Debian/Ubuntu AFAIK use their own `mkinitramfs` tool that doesn't utilize systemd at all.

So sorry, this is a distro-specific thing. I mean, different distros also name their kernels differently, and they patch grub to detect those kernels. This stuff just doesn't belong in systemd.

**Comment by eugenesan** [comment]
@YHNdnzj 
Thanks for responding.

I am not familiar with systemd codebase and probably missing something, but looking at the code: https://github.com/systemd/systemd/blob/3b8b6910d180fbb4b6299885617ac1114d01be9a/src/kernel-install/90-loaderentry.install.in#L74 I don't see an attempt to figure appropriate boot options.
Or, on other systems with different "mkinitrd" one of the `/.*/.*/cmdline` is being generated by something?
On Ubuntu/Debian only /proc/cmdline is available (obsessively).

If there is a working model of "mkinitrd" (or something else) passing boot mount options to sd-boot, could you describe it or point me in direction?
Also if there is a relevant code in systemd, could you point me to it?

**Comment by YHNdnzj** [comment]
>  I don't see an attempt to figure appropriate boot options.

As I said, there's no such thing as "appropriate boot options". It's entirely up to the distros/users to decide how to name your DM devices/which keys to use. I'm an Arch user, and Arch users always fills in the kernel cmdline options on their own. Therefore I might be quite ill-informed about how other distros handle this. But my point still stands, that there's no generic way to "find out boot options". The relevant machinery belongs to the distro, rather than systemd.

cc @bluca @keszybz: Could you explain how this is handled in Debian/Fedora?

**Comment by DaanDeMeyer** [comment]
dracut reads /etc/kernel/cmdline (and maybe dropins?) and falls back to /proc/cmdline

**Comment by YHNdnzj** [comment]
> dracut reads /etc/kernel/cmdline (and maybe dropins?) and falls back to /proc/cmdline

Hmm, so I guess the Fedora installer just writes the info directly to the file and never looks back? I.e. only Debian has the complex auto-detection thing?

**Comment by eugenesan** [comment]
I've opened an issue on Debian Kernel Team (https://salsa.debian.org/kernel-team/kernel-team/-/issues/3).
Let's see what they have to say on subject.

**Comment by poettering** [comment]
OK, let's close this here as this doesn't look like it has anything to do with systemd, and is a distro/initrd generator issue.
